nixpkgs/nixpkgs.bzl

```diff
@@ -925,6 +925,41 @@ pkgs.runCommand "bazel-nixpkgs-java-runtime"
   ''
 """
 
+def _nixpkgs_java_toolchain_impl(repository_ctx):
+    cpu = get_cpu_value(repository_ctx)
+    exec_constraints, target_constraints = ensure_constraints(repository_ctx)
+
+    repository_ctx.file(
+        "BUILD.bazel",
+        executable = False,
+        content = """\
+package(default_visibility = ["//visibility:public"])
+
+toolchain(
+    name = "java-toolchain-{cpu}",
+    toolchain = "{runtime}",
+    toolchain_type = "@bazel_tools//tools/jdk:runtime_toolchain_type",
+    exec_compatible_with = {exec_constraints},
+    target_compatible_with = {target_constraints},
+)
```

Comment:
I was going to suggest to use [`local_java_runtime`](https://github.com/bazelbuild/bazel/blob/41feb616ae18e21fdba3868e4c298b0b83012f10/tools/jdk/local_java_repository.bzl#L41) with `java_home = "", runtime_name = "{runtime}"` following the documentation [here](https://docs.bazel.build/versions/main/bazel-and-java.html#configuring-additional-execution-toolchains). However, I see that `local_java_runtime` doesn't expose `exec|target_compatible_with`, so this doesn't work for us.

I see that `local_java_runtime` generates a couple [`config_setting`s](https://github.com/bazelbuild/bazel/blob/41feb616ae18e21fdba3868e4c298b0b83012f10/tools/jdk/local_java_repository.bzl#L67-L90) to then assign to the `toolchain`'s [`target_setting`](https://docs.bazel.build/versions/main/be/platform.html#toolchain). I haven't followed all the bread crumbs, but I suspect this is needed for the `--java_runtime_version` flag to work.

Perhaps a simple fix is to copy that code to generate those `config_setting`s here as well and raise a ticket on `bazelbuild/bazel` to expose `exec|target_compatible_with` on `local_java_runtime`, so that we can re-use it here in future.

Reply:
good idea, I’ll try to get to this tomorrow.

Reply:
Tomorrow turned out to be today …

I added the config settings and this does indeed make `--java_runtime_version` work. However, I noticed that `local_java_runtime` also sets up a bunch of default_java_toolchain rules. I’m not entirely sure what this is for but I think one is for compilation while the other is for test execution?

I wonder if duplicating all of this is really a good idea. What do you think of using `local_java_runtime` and accept that it does not set the `target_compatible_with` and `exec_compatible_with`. To differentiate between nix and non-nix platforms you can use `--java_runtime_version` which while being less automatic is also not that terrible and I think at least for the Daml repo would work very well.

Reply:
I pushed a commit for that. Happy to revert if you think that’s a bad idea.

Reply:
> I added the config settings and this does indeed make `--java_runtime_version` work.

Good to know, thanks for adding and testing!

> However, I noticed that local_java_runtime also sets up a bunch of default_java_toolchain rules. I’m not entirely sure what this is for but I think one is for compilation while the other is for test execution?

Good catch, yes, runtime is for execution (tests and other kinds) and toolchain for compilation.

> To differentiate between nix and non-nix platforms you can use --java_runtime_version which while being less automatic is also not that terrible and I think at least for the Daml repo would work very well.

Are they different? I would have thought, e.g. in the Daml setup, that the versions are the same on Windows and Unix.

Reply:
The versions are the same but the toolchains are not. Currently the exec and target constraints make sure we select the right one. If we use `local_java_runtime` we need another option to pick the right toolchain depending on the platform. Bazel sets config constraints for `java_runtime_version` for both the name and the version. if you look at the test case. You can then specify either `--java_runtime_version=name`, `--java_runtime_version=version` or `--java_runtime_version=name_version`. So my suggestion is that for the daml repo we would set `build:linux --java_runtime_version=nixpkgs_java` which matches on the name and similar for macos while on Windows, we pick the name passed there which is different.

This isn’t as nice, we really should be able to just set the constraints on the platform. But it seems preferable to me than trying to duplicate both `local_java_runtime` and `default_java_toolchain`.

Reply:
Ah, I see. Yes, that should work. Indeed, it's not as nice as the platform being sufficient. But, fixing that upstream by exposing those `_compatible_with` attributes seems preferable.